File: src/youtube-api/services/youtube-playlist-api.service.ts
import { GaxiosResponse } from "gaxios";

import { OAuth2Client } from "googleapis-common";

import { Youtube, Schema$Playlist } from "../types/youtube.aliases";

import { YOUTUBE_PLAYLIST_PRIVACY_STATUS } from "../metadata/youtube-playlist-privacy-status.enum";

export class YoutubePlaylistApiService {
  constructor(private client: OAuth2Client, private youtube: Youtube) {}

  public getPlaylists(): Promise<GaxiosResponse<Schema$Playlist>[]> {
    return new Promise(async (resolve, reject) => {
      this.youtube.playlists
        .list({
          auth: this.client,
          mine: true
        })
        .then(async firstPage => {
          if (firstPage.data.nextPageToken) {
            return resolve([
              firstPage,
              ...(await this.getNextPlaylistPages(firstPage.data.nextPageToken))
            ]);
          }

          return resolve([firstPage]);
        })
        .catch(error => reject(error));
    });
  }

  private getNextPlaylistPages(
    pageToken: string
  ): Promise<GaxiosResponse<Schema$Playlist>[]> {
    return new Promise(async (resolve, reject) => {
      const pages = [
        await this.youtube.playlists.list({
          auth: this.client,
          mine: true,
          pageToken: pageToken
        })
      ];
      const nextPageToken = pages[0]?.data?.nextPageToken;

      if (nextPageToken) {
        pages.push(...(await this.getNextPlaylistPages(nextPageToken)));
      }

      resolve(pages);
    });
  }

  public addPlaylist(
    title: string,
    description: string,
    privacy: YOUTUBE_PLAYLIST_PRIVACY_STATUS
  ): Promise<GaxiosResponse<Schema$Playlist>> {
    return this.youtube.playlists.insert({
      auth: this.client,
      part: "snippet,status",
      requestBody: {
        snippet: {
          title: title,
          description: description,
          defaultLanguage: "en"
        },
        status: {
          privacyStatus: privacy
        }
      }
    });
  }
}

File: packages/youtube-api/lib/src/services/youtube_playlist_api.service.dart
import 'package:googleapis/youtube/v3.dart';

class YoutubePlaylistApiService {
  final YoutubeApi _youtube;

  YoutubePlaylistApiService(this._youtube);

  Future<List<PlaylistListResponse>> getPlaylists() {
    return _youtube.playlists.list("", mine: true).then((firstPage) async {
      if (firstPage.nextPageToken != null) {
        return [
          firstPage,
          ...(await _getNextPlaylistPages(firstPage.nextPageToken))
        ];
      }

      return [firstPage];
    });
  }

  Future<List<PlaylistListResponse>> _getNextPlaylistPages(
      String pageToken) async {
    final pages = [
      await _youtube.playlists.list("", mine: true, pageToken: pageToken)
    ];

    final nextPageToken = pages[0]?.nextPageToken;

    if (nextPageToken != null) {
      pages.addAll(await _getNextPlaylistPages(nextPageToken));
    }

    return pages;
  }

  Future<Playlist> addPlaylist(
      String title, String description, String privacy) {
    final request = Playlist.fromJson({
      "playlistSnippet": PlaylistSnippet.fromJson(
          {"title": title, "description": description}),
      "status": PlaylistStatus.fromJson({"privacyStatus": privacy})
    });

    return _youtube.playlists.insert(request, "snippet,status");
  }
}
